File: Example/QDNavigationBar/QDConfigEditor/QDConfigColorView.m
//
//  QDConfigColorView.m
//  QDCustomNavigation_Example
//
//  Created by Sinno on 2020/8/5.
//  Copyright © 2020 sinno93. All rights reserved.
//

#import "QDConfigColorView.h"
#import <Masonry/Masonry.h>
#import "QDColorPickerController.h"


@interface QDConfigColorView ()
@property (nonatomic, strong) UILabel *titleLabel;
@property (nonatomic, strong) UIView *colorView;
@end

@implementation QDConfigColorView

#pragma mark - Initialization
// 初始化方法
- (instancetype)initWithFrame:(CGRect)frame {
    if (self = [super initWithFrame:frame]) {
        [self configSubviews];
    }
    return self;
}

#pragma mark - Super methods

#pragma mark - Action Methods

- (void)colorTap:(UITapGestureRecognizer *)tap {
    QDColorPickerController *vc = [[QDColorPickerController alloc] init];
    vc.color = self.color;
    
    [[self getViewController] presentViewController:vc animated:NO completion:nil];
    __weak __typeof(self)weakSelf = self;
    vc.colorChanged = ^(UIColor * _Nonnull color) {
        weakSelf.color = color;
        if (weakSelf.colorChanged) {
            weakSelf.colorChanged(color);
        }
    };
}

#pragma mark - Delegate


#pragma mark - Public Methods
// 公开的方法
- (void)setTitle:(NSString *)title {
    self.titleLabel.text = title;
}

- (NSString *)title {
    return self.titleLabel.text;
}

- (void)setColor:(UIColor *)color {
    self.colorView.backgroundColor = color;
    CGFloat red = 0;
    CGFloat green = 0;
    CGFloat blue = 0;
    CGFloat alpha = 0;
    [color getRed:&red green:&green blue:&blue alpha:&alpha];
    
    UIColor *contrastColor = nil;
    CGFloat grayLevel = (0.299 * red + 0.587 * green + 0.114 * blue);
    if (grayLevel > 0.5 || alpha<0.5) {
        contrastColor = UIColor.blackColor;
    } else {
        contrastColor = UIColor.whiteColor;
    }
    _colorView.tintColor = contrastColor;
}

- (UIColor *)color {
    return self.colorView.backgroundColor;
}
#pragma mark - Private Methods
// 私有的方法
- (void)configSubviews {
    [self addSubview:self.titleLabel];
    [self addSubview:self.colorView];

    [self.titleLabel mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerY.equalTo(self);
        make.leading.equalTo(self).offset(15);
    }];
    
    [self.colorView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerY.equalTo(self);
        make.trailing.equalTo(self).offset(-25);
        make.size.mas_equalTo(CGSizeMake(40, 40));
    }];
}

- (UIViewController *)getViewController {
    UIResponder *responder = self;
    while (responder.nextResponder) {
        if ([responder.nextResponder isKindOfClass:UIViewController.class]) {
            return (UIViewController *)responder.nextResponder;
        }
        responder = responder.nextResponder;
    }
    return nil;
}
#pragma mark - Getter && Setter
// Getter和Setter方法
- (UILabel *)titleLabel {
    if (!_titleLabel) {
        UILabel *label = [[UILabel alloc] init];
        label.font = [UIFont systemFontOfSize:20.0];
        label.textColor = UIColor.blackColor;
        _titleLabel = label;
    }
    return _titleLabel;
}

- (UIView *)colorView {
    if (!_colorView) {
        
        _colorView = [[UIView alloc] init];
        _colorView.layer.borderColor = UIColor.grayColor.CGColor;
        _colorView.layer.borderWidth = 1.0;
        UITapGestureRecognizer *tapGesture = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(colorTap:)];
        [_colorView addGestureRecognizer:tapGesture];
        UIImageView *imageView = [[UIImageView alloc] init];
        imageView.image = [[UIImage imageNamed:@"expand_icon"] imageWithRenderingMode:UIImageRenderingModeAlwaysTemplate];
        
        [_colorView addSubview:imageView];
        [imageView mas_makeConstraints:^(MASConstraintMaker *make) {
            make.center.equalTo(_colorView);
            make.size.mas_equalTo(CGSizeMake(20, 20));
        }];
    }
    return _colorView;
}
@end

File: Example/QDNavigationBar/QDConfigEditor/QDConfigColorView.swift
//
//  QDConfigColorView.swift
//  QDNavigationBar_Example
//
//  Created by zqd on 2024/6/5.
//  Copyright © 2024 CocoaPods. All rights reserved.
//

import UIKit
import SnapKit

class QDConfigColorView: UIView {
    private lazy var titleLabel: UILabel = {
        let label = UILabel()
        label.font = UIFont.systemFont(ofSize: 20.0)
        label.textColor = .black
        return label
    }()
    
    private lazy var colorView: UIView = {
        let view = UIView()
        view.layer.borderColor = UIColor.gray.cgColor
        view.layer.borderWidth = 1.0
        
        let tapGesture = UITapGestureRecognizer(target: self, action: #selector(colorTap(_:)))
        view.addGestureRecognizer(tapGesture)
        
        let imageView = UIImageView()
        imageView.image = UIImage(named: "expand_icon")?.withRenderingMode(.alwaysTemplate)
        view.addSubview(imageView)
        imageView.snp.makeConstraints { make in
            make.center.equalTo(view)
            make.size.equalTo(CGSize(width: 20, height: 20))
        }
        
        return view
    }()
    
    private lazy var bgImageView: UIImageView = {
        let imageView = UIImageView()
        imageView.image = UIImage(named: "alphaImage")
        return imageView
    }()
    
    var color: UIColor = .white {
        didSet {
            colorView.backgroundColor = color
            updateTintColor()
        }
    }
    
    var title: String {
        get {
            titleLabel.text ?? ""
        }
        
        set {
            titleLabel.text = newValue
        }
    }
    
    var colorChanged: ((UIColor) -> Void)?

    // MARK: - Initialization
    override init(frame: CGRect) {
        super.init(frame: frame)
        configSubviews()
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
        configSubviews()
    }
    
    // MARK: - Actions
    @objc private func colorTap(_ tap: UITapGestureRecognizer) {
        let vc = QDColorPickerController()
        vc.color = self.color
        
        getViewController()?.present(vc, animated: false, completion: nil)
        vc.colorChanged = { [weak self] color in
            self?.color = color
            self?.colorChanged?(color)
        }
    }
    
    // MARK: - Private Methods
    private func configSubviews() {
        addSubview(titleLabel)
        addSubview(bgImageView)
        addSubview(colorView)
        
        titleLabel.snp.makeConstraints { make in
            make.centerY.equalTo(self)
            make.leading.equalTo(self).offset(15)
        }
        
        colorView.snp.makeConstraints { make in
            make.centerY.equalTo(self)
            make.trailing.equalTo(self).offset(-25)
            make.size.equalTo(CGSize(width: 40, height: 40))
        }
        
        bgImageView.snp.makeConstraints { make in
            make.edges.equalTo(colorView)
        }
    }
    
    private func updateTintColor() {
        var red: CGFloat = 0
        var green: CGFloat = 0
        var blue: CGFloat = 0
        var alpha: CGFloat = 0
        color.getRed(&red, green: &green, blue: &blue, alpha: &alpha)
        
        let grayLevel = (0.299 * red + 0.587 * green + 0.114 * blue)
        colorView.tintColor = (grayLevel > 0.5 || alpha < 0.5) ? .black : .white
    }
    
    private func getViewController() -> UIViewController? {
        var responder: UIResponder? = self
        while let nextResponder = responder?.next {
            if let viewController = nextResponder as? UIViewController {
                return viewController
            }
            responder = nextResponder
        }
        return nil
    }
    
}

